### DISCUSSION TITLE
Not loading environment correctly

### DISCUSSION THREAD

**Discussion by ttunguz** [Taitava/obsidian-shellcommands#357] [Debugging and testing]
Hi, I'm using this great extension to pass a url to a ruby script which will then modify the contents of the page by calling a few different functions. 

My challenge is that despite having all the ruby libraries in the path (I've created another script which echoes the Path, & it's consistent with my development environment), when I run the code, I receive a cannot load such file in the browser.

Is shellcommands instantiating a full terminal environment in the background? Could you foresee any reason why a library wouldn't be visible to this environment? Thank you.

**Comment by ttunguz** [comment]
I found it was idiosyncratic to my environment. Thanks!

**Reply by Taitava** [reply]
Hi! 🙂 So you got it working? Or need still help? 🙂

**Comment by ttunguz** [comment]
I still need help! I'm using the command to call a ruby script.

Importing the system gems works fine but requiring something like nokogiri
gem throws an error that the environment can't find it. I double checked
the path & it includes the path to the gems. I don't know what to do.

**Comment by ttunguz** [comment]
steps to reproduce :

0. install ruby, & the gem nokogiri via gem install nokogiri
1. create a ruby script called test.rb

<<<
require 'nokogiri'
require 'date'

p 'this is a test'

<<<<

2.  call the script from obsidian
3. see the error message that nokogiri could not be found

**Reply by Taitava** [reply]
Sorry, I don't know Ruby, and I need to limit the support I can give by not installing programs that I don't have expertise with 🙂. If there are any Ruby experts reading this conversation, please let us know if you can help with this problem. 🙂 

> Is shellcommands instantiating a full terminal environment in the background? Could you foresee any reason why a library wouldn't be visible to this environment?

What do you mean by a _full terminal environment_? If you mean, does the plugin invoke a _shell_ when executing commands, then yes, it does, and so syntax that works e.g. in `Bash`/`Dash`/`Zsh` also works in this plugin. However, what this plugin does miss by default, is additions to the `$PATH` environment variable, that shells normally tend to do automatically.

In case the Ruby libraries are missing due to the `$PATH` environment variable, you can try these things:
A) If you know the directories that need to be added to `$PATH`, [please follow these instructions to define those additions in the plugin's settings](https://publish.obsidian.md/shellcommands/Environments/Additions+to+the+PATH+environment+variable#An+easier+way+to+add+directories+to+%60PATH%60). This might be easier than the section B, although B gives you better abilities to customise the shell invocation.
B) Alternatively, you could [define a custom shell](https://publish.obsidian.md/shellcommands/Environments/Custom+shells/Custom+shells). This allows you to declare arguments that will be used when invoking the shell. E.g. for `Bash` shell, you can [add the `--login` shell argument, which makes Bash read certain profile files](https://www.man7.org/linux/man-pages/man1/bash.1.html#INVOCATION) before executing your shell command(s). This should fix the `$PATH` problem without you needing to explicitly define the directories like instructed in A. Here I'm just assuming that you are using Bash. But if you are using some other shell, please let me know and I can refine my instructions.

When using a non-custom shell, the plugin only passes [`-c {{!shell_command_content}}` when invoking the shell (or `/d /s /c {{!shell_command_content}}` if the shell is CMD.EXE)](https://nodejs.org/api/child_process.html#shell-requirements), so it's missing the `--login` argument by default, which explains why `$PATH` is often missing directories.

Was this of help at all? 🙂

**Reply by Taitava** [reply]
If you still encounter problems, please include a screenshot of the error message, so that I can see the exact error message and the situation it occurs in. 👍

**Comment by ttunguz** [comment]
Thank you. I rearchitected this to solve the environment variable problems by moving it to a cloud service.

I appreciate your help!
On 2023-06-27, Jarkko Linnanvirta wrote:
>If you still encounter problems, please include a screenshot of the error message, so that I can see the exact error message and the situation it occurs in. 👍
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/Taitava/obsidian-shellcommands/discussions/357#discussioncomment-6296352
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

-- 
Tom Tunguz
tomtunguz.com
